core/chains/evm/txmgr/evm_tx_store.go

```diff
@@ -1767,6 +1768,68 @@ func (o *evmTxStore) Abandon(ctx context.Context, chainID *big.Int, addr common.
+	err := qq.Select(&dbEtxs, sql, chainID.String(), metaValue, pq.Array(states))
+	txes := make([]*Tx, len(dbEtxs))
+	dbEthTxsToEvmEthTxPtrs(dbEtxs, txes)
+	return txes, pkgerrors.Wrap(err, "failed to FindTxesByMetaFieldAndStates")
+}
+
+// Find transactions with a non-null TxMeta field that was provided by transaction states
+func (o *evmTxStore) FindTxesWithMetaFieldByStates(ctx context.Context, metaField string, states []txmgrtypes.TxState, chainID *big.Int) (txes []*Tx, err error) {
+	var cancel context.CancelFunc
+	ctx, cancel = o.mergeContexts(ctx)
+	defer cancel()
+	qq := o.q.WithOpts(pg.WithParentCtx(ctx))
+	var dbEtxs []DbEthTx
+	sql := fmt.Sprintf("SELECT * FROM evm.txes WHERE meta->'%s' IS NOT NULL AND state = ANY($1) AND evm_chain_id = $2", metaField)
+	err = qq.Select(&dbEtxs, sql, pq.Array(states), chainID.String())
+	txes = make([]*Tx, len(dbEtxs))
+	dbEthTxsToEvmEthTxPtrs(dbEtxs, txes)
+	return txes, pkgerrors.Wrap(err, "failed to FindTxesWithMetaFieldByStates")
+}
+
+// Find transactions with a non-null TxMeta field that was provided and a receipt block number greater than or equal to the one provided
+func (o *evmTxStore) FindTxesWithMetaFieldByReceiptBlockNum(ctx context.Context, metaField string, blockNum int64, chainID *big.Int) (txes []*Tx, err error) {
+	var cancel context.CancelFunc
+	ctx, cancel = o.mergeContexts(ctx)
+	defer cancel()
+	qq := o.q.WithOpts(pg.WithParentCtx(ctx))
+	var dbEtxs []DbEthTx
+	sql := fmt.Sprintf("SELECT et.* FROM evm.txes et JOIN evm.tx_attempts eta on et.id = eta.eth_tx_id JOIN evm.receipts er on eta.hash = er.tx_hash WHERE et.meta->'%s' IS NOT NULL AND er.block_number >= $1 AND et.evm_chain_id = $2", metaField)
+	err = qq.Select(&dbEtxs, sql, blockNum, chainID.String())
+	txes = make([]*Tx, len(dbEtxs))
+	dbEthTxsToEvmEthTxPtrs(dbEtxs, txes)
+	return txes, pkgerrors.Wrap(err, "failed to FindTxesWithMetaFieldByReceiptBlockNum")
+}
+
+// Find transactions loaded with transaction attempts and receipts by transaction IDs and states
+func (o *evmTxStore) FindTxesWithAttemptsAndReceiptsByIdsAndState(ctx context.Context, ids []big.Int, states []txmgrtypes.TxState, chainID *big.Int) (txes []*Tx, err error) {
+	err = o.q.Transaction(func(tx pg.Queryer) error {
+		var dbEtxs []DbEthTx
+		if err = tx.Select(&dbEtxs, `SELECT * FROM evm.txes WHERE id = ANY($1) AND state = ANY($2) AND evm_chain_id = $3`, pq.Array(ids), pq.Array(states), chainID.String()); err != nil {
+			return pkgerrors.Wrapf(err, "failed to find evm.txes")
+		}
+		txes = make([]*Tx, len(dbEtxs))
+		dbEthTxsToEvmEthTxPtrs(dbEtxs, txes)
+		if err = o.LoadTxesAttempts(txes, pg.WithQueryer(tx)); err != nil {
+			return pkgerrors.Wrapf(err, "failed to load evm.tx_attempts for evm.tx")
+		}
+		if err = loadEthTxesAttemptsReceipts(tx, txes); err != nil {
+			return pkgerrors.Wrapf(err, "failed to load evm.receipts for evm.tx")
+		}
+		return nil
```

Comment:
looking at the underylying implementation, each tx can have n attempts and each attempt can have n receipts. Is that the correct understanding? What are scenarios when an attempt can have multiple receipts? Would be great to have these details as comments too.

Reply:
Each tx can have n attempts but I believe there should only be one receipt per tx. I'm trying to track down the reason why the struct supports more than one. I'll write whatever I find into comments

Reply:
I've updated the comments in the `loadEthTxesAttemptsReceipts` method to address this. Looks like the expectation is to only have one receipt per tx. We delete any existing receipts prior to handling reorgs. I didn't get any concrete context on why we support a list of receipts so I didn't add it in the comment but the guess is we may have wanted to keep previous receipts in case of reorgs to maintain history in case of post-mortem investigations.